### PR TITLE
subsys: power: rename CPU_LPS_n power states

### PR DESCRIPTION
**Pull request by mnkp**
This PR is split into two commits:

**power: simplify SYS_POWER_*_SUPPORTED Kconfig options**
This commit removes dependency on SYS_POWER_LOW_POWER_STATES_SUPPORTED, SYS_POWER_DEEP_SLEEP_STATES_SUPPORTED Kconfig options. Power management SYS_POWER_LOW_POWER_STATES, SYS_POWER_DEEP_SLEEP_STATES options depend now directly on specific power states supported by the given SoC. This simplifies maintenance of SoC Kconfig files.

**power: rename CPU_LPS_n power states**
CPU_LPS_n name used to indicate a low power state is cryptic and incorrect. The low power states act on the whole SoC and not exclusively on the CPU. This patch renames CPU_LPS_n states to LOW_POWER_n. Also HAS_ pattern for Kconfig options is used in favor of a non standard _SUPPORTED. Naming of deep sleep states was adjusted accordingly.

Following is a detailed list of string replacements used:
```
s/SYS_POWER_STATE_CPU_LPS_(\d)_SUPPORTED/HAS_STATE_LOW_POWER_$1/
s/SYS_POWER_STATE_CPU_LPS_(\d)/SYS_POWER_STATE_LOW_POWER_$1/
s/SYS_POWER_STATE_DEEP_SLEEP_(\d)_SUPPORTED/HAS_STATE_DEEP_SLEEP_$1/
```
This is continuation of work done in #13317.

### CODE REVIEW

File: arch/Kconfig [hunk 1]
```diff
@@ -347,56 +347,56 @@ config ARCH_HAS_THREAD_ABORT
...
-config SYS_POWER_LOW_POWER_STATES_SUPPORTED
+config SYS_POWER_HAS_LOW_POWER_STATES
 	# Hidden
 	bool
 	help
 	  This option signifies that the target supports the SYS_POWER_LOW_POWER_STATES
 	  configuration option.
 
-config SYS_POWER_STATE_CPU_LPS_1_SUPPORTED
+config SYS_POWER_STATE_LOW_POWER_1_SUPPORTED
```
Comment (nashif) [inline arch/Kconfig]: shouldn't this be SYS_POWER_HAS_LOW_POWER_1 or something lie that?

Comment (mnkp) [reply arch/Kconfig]: Sorry, fixed.

**Comment by pizi-nordic** [comment]
Looks good, but one idea popped in my mind:

SYS_POWER_HAS.... suggests that system power has something. Maybe SOC_HAS... or just HAS_... would be better?

**Comment by mnkp** [comment]
> SYS_POWER_HAS.... suggests that system power has something, which is not true.
> Maybe SOC_HAS... or just HAS_... would be better?

Makes sense, both option look good. Anyone, any preference?

**Comment by mnkp** [comment]
There is one more naming issue I would like to discuss. Currently the description of `sys_pm_notify_lps_entry()` function says:
```
/**
 * @brief Application defined function for Lower Power entry
 *
 * Application defined function for doing any target specific operations
 * for low power entry.
 */
```
The issue here is that the function will be executed when entering `SYS_POWER_STATE_LOW_POWER_n` as well as `SYS_POWER_STATE_DEEP_SLEEP_n`. The description suggests something else. The naming convention for low power states, namely `STATE_LOW_POWER_n`, `STATE_DEEP_SLEEP_n` makes it very difficult to distinguish between any low power state and `STATE_LOW_POWER_n` specifically. `sys_pm_notify_lps_entry()` function is just an example, there are several other places in the API where we have the same problem and the description is currently ambiguous.

One solution would be to rename `SYS_POWER_STATE_LOW_POWER_n` to `SYS_POWER_STATE_SLEEP_n`. So we would have 'Sleep' low power modes and 'Deep Sleep' low power modes. But maybe there is a better suggestion.

As a general comment renaming operation is not very expensive so if anyone has any proposals on how to improve naming implementing the change is not a problem.

**Comment by nashif** [comment]
> > SYS_POWER_HAS.... suggests that system power has something, which is not true.
> > Maybe SOC_HAS... or just HAS_... would be better?
> 
> Makes sense, both option look good. Anyone, any preference?

HAS_.... works for me.

**Review by dbkinder — DISMISSED** [review]
+1 for doc changes

**Comment by pizi-nordic** [comment]
Maybe we use "power_state" instead of "low_power_state" in the names?

**Comment by mnkp** [comment]
> Maybe we use "power_state" instead of "low_power_state" in the names?

You mean, like changing the `sys_pm_notify_lps_entry()` function name to `sys_pm_notify_power_state_entry()`. And changing the description to "Application defined function for Power State entry"?

**Comment by pizi-nordic** [comment]
Yes. But the docs should emphasise that this is a notification, not implementation.

**Comment by mnkp** [comment]
I've added a commit "power: doc: fix usage of 'low power state' phrase" which highlights places where the usage of 'low power state' phrase in the documentation is ambiguous. In all the cases where the change was done we mean 'low power or deep sleep state'.

I still think we should drop 'low power state' name and use 'sleep state' instead. So we would have
- active state (normal operation)
- sleep state (low power state where we don't lose execution context)
- deep sleep state (low power state where we lose execution context)

The name 'sleep state' makes it clear that the CPU is not running. But I would need some OKs before I do the change.

I still need to prepare one more commit which aligns function names (I may do it also as a separate PR). Most of the functions use currently prefix `sys_pm_` I was thinking of renaming them to `sys_power_`. Any better recommendation? Should I keep the old prefix?

**Comment by pizi-nordic** [comment]
Since you touched prefixes:

I am trying to model power states in a similar way as it is done in ACPI. This means that there will be several components in the system with own power level as well as one global power state which will map to power states of the components. Example:

Global Power State 0:
- Device 1: Active
- Device 2: Active or Suspended
- Device 3: Suspended
- SoC: Active or in LPS_1

Global Power State 1:
- Device 1: Active or Suspended
- Device 2: OFF
- Device 3: Active
- SoC: Active or in [LSP1, LSP2, LPS_3]

Global Power State 2:
- Device 1: OFF
- Device 2: OFF
- Device 3: Active
- SoC: Deep sleep.

At the moment we have "system" power management which in fact is a "SoC" power management. I planned to rename it to "soc_pm", but since you touched the topic, we might try to implement desired naming structure now.

**Comment by pizi-nordic** [comment]
+1 for removing "low power" state and use naming proposed by @mnkp.

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, thanks.

**Review by nashif — APPROVED** [review]

**Pull request merged**